nexus/reconfigurator/execution/src/datasets.rs

```diff
@@ -4,97 +4,218 @@
 
 //! Ensures dataset records required by a given blueprint
 
+use crate::Sled;
+
+use anyhow::anyhow;
 use anyhow::Context;
+use futures::stream;
+use futures::StreamExt;
 use nexus_db_model::Dataset;
 use nexus_db_queries::context::OpContext;
 use nexus_db_queries::db::DataStore;
-use nexus_types::deployment::BlueprintZoneConfig;
-use nexus_types::deployment::DurableDataset;
+use nexus_types::deployment::BlueprintDatasetConfig;
+use nexus_types::deployment::BlueprintDatasetDisposition;
+use nexus_types::deployment::BlueprintDatasetsConfig;
 use nexus_types::identity::Asset;
+use omicron_common::disk::DatasetConfig;
+use omicron_common::disk::DatasetsConfig;
+use omicron_uuid_kinds::DatasetUuid;
 use omicron_uuid_kinds::GenericUuid;
-use omicron_uuid_kinds::OmicronZoneUuid;
+use omicron_uuid_kinds::SledUuid;
 use slog::info;
+use slog::o;
 use slog::warn;
-use std::collections::BTreeSet;
+use std::collections::BTreeMap;
+
+/// Idempotently ensures that the specified datasets are deployed to the
+/// corresponding sleds
+pub(crate) async fn deploy_datasets(
+    opctx: &OpContext,
+    sleds_by_id: &BTreeMap<SledUuid, Sled>,
+    sled_configs: &BTreeMap<SledUuid, BlueprintDatasetsConfig>,
+) -> Result<(), Vec<anyhow::Error>> {
+    let errors: Vec<_> = stream::iter(sled_configs)
+        .filter_map(|(sled_id, config)| async move {
+            let log = opctx.log.new(o!(
+                "sled_id" => sled_id.to_string(),
+                "generation" => config.generation.to_string(),
+            ));
+
+            let db_sled = match sleds_by_id.get(&sled_id) {
+                Some(sled) => sled,
+                None => {
+                    let err = anyhow!("sled not found in db list: {}", sled_id);
+                    warn!(log, "{err:#}");
+                    return Some(err);
+                }
+            };
+
+            let client = nexus_networking::sled_client_from_address(
+                sled_id.into_untyped_uuid(),
+                db_sled.sled_agent_address,
+                &log,
+            );
+
+            let config: DatasetsConfig = match config.clone().try_into() {
+                Ok(config) => config,
+                Err(err) => return Some(err)
```

Comment:
Do you want to `warn!()` this error too? (Looks like the other errors branches all get it)

Reply:
With the removal of a stringified CompressionAlgorithm, this actually doesn't need to return an error anymore. Updated.